File: R/forceScheme.R
#' Force Scheme Projection
#'
#' Creates a 2D representation of the data based on a dissimilarity matrix. A few
#' modifications have been made in relation to the method described in the
#' literature: shuffled indices are used to minimize the order dependency
#' factor, only a fraction of delta is used for better stability and a tolerance
#' factor was introduced as a second stop criterion.
#'
#' @param d A dissimilarity structure such as that returned by dist or a full
#'   symmetric matrix containing the dissimilarities.
#' @param initial A initial 2D configuration (optional). A random configuration
#'   will be created when omitted.
#' @param max.it The maximum number of iterations that the method will run.
#' @param tol The tolerance for the accumulated error between iterations. Set it
#'   to 0 to guarantee it will run max.it times.
#' @param verbose A flag that indicates if the progress should be printed.
#' @return The 2D representation of the data.
#'
#' @references Eduardo Tejada, Rosane Minghim, Luis Gustavo Nonato: On improved
#'   projection techniques to support visual exploration of multi-dimensional
#'   data sets. Information Visualization 2(4): 218-231 (2003)
#'
#' @examples
#' # Eurodist example
#' proj = forceScheme(eurodist)
#' plot(proj, type = "n", xlab ="", ylab ="", asp=1, axes=FALSE, main="")
#' text(proj, labels(eurodist), cex = 0.6)
#'
#' # Iris example
#' proj = forceScheme(dist(iris[,1:4]))
#' plot(proj, col=iris$Species)
#' @seealso \code{\link[stats]{dist}} (stats) and \code{\link[proxy]{dist}}
#'   (proxy) for d computation
#'
#' @useDynLib mp
#' @export
forceScheme = function(d, initial=NULL, max.it=50, tol=0.1, verbose=F) {
  EPSILON = 1E-5 # minimum distance between points
  fraction = 8.0 # fraction of delta

  # convert d to a matrix
  dmat = as.matrix(d)

  # get the number of elements
  n = nrow(dmat)

  # create a random initial configuration if none was given
  if (is.null(initial)) {
    initial = matrix(runif(n*2,0,1), ncol=2)
  }

  # check initial configuration
  if (nrow(initial) != n) {
    stop("Provided initial configuration does not match the input matrix length");
  }

  p = initial

  # switch core implementation
  pC = .C("force_scheme",
          p = as.numeric(p),
          dmat = as.numeric(dmat),
          n = as.integer(n),
          max_it = as.integer(max.it),
          tol = as.double(tol),
          EPSILON = as.double(EPSILON),
          fraction = as.double(fraction))$p

  p = cbind(pC[1:n], pC[(n+1):(2*n)])

  # return the 2D rojection
  return(p)
}

File: src/forceScheme.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

/*
 * Arrange the N elements of ARRAY in random order.
 * Only effective if N is much smaller than RAND_MAX;
 * if this may not be the case, use a better random
 * number generator.
 */
static void shuffle(size_t *array, size_t n)
{
    if (n > 1) {
        size_t i;
        for (i = 0; i < n - 1; i++) {
            size_t j = i + rand() / (RAND_MAX / (n - i) + 1);
            int t = array[j];
            array[j] = array[i];
            array[i] = t;
        }
    }
}

void force_scheme(double p[],
        const double dmat[],
        const int *n,
        const int *max_it,
        const double *tol,
        const double *EPSILON,
        const double *fraction)
{
    double prev_delta_sum = 1. / 0., delta_sum, d2, dn, delta, diff_x, diff_y;
    size_t i, j, k, p1_index, p2_index;
    size_t size = (size_t) *n;

    size_t s_j[size], s_k[size];
    for (i = 0; i < size; i++)
        s_j[i] = s_k[i] = i;

    int asd = 0;

    for (i = 0; i < *max_it; i++) {
        shuffle(s_j, size);
        for (j = 0; j < size; j++) {
            p1_index = s_j[j];

            delta_sum = 0;
            shuffle(s_k, size);
            for (k = 0; k < size; k++) {
                p2_index = s_k[k];

                if (p1_index == p2_index)
                    continue;

                diff_x = p[p2_index] - p[p1_index];
                diff_y = p[p2_index + size] - p[p1_index + size];

                // TODO: Verify alternatives to avoid over/underflow
                d2 = sqrt(diff_x * diff_x + diff_y * diff_y);

                if (d2 < *EPSILON)
                    d2 = *EPSILON;

                dn = dmat[p1_index * size + p2_index];
                delta = (dn - d2) / *fraction;
                delta_sum += fabs(delta);
                p[p2_index]        += delta * (diff_x / d2);
                p[p2_index + size] += delta * (diff_y / d2);
            }

        }

        if (fabs(prev_delta_sum - delta_sum) < *tol)
            break;
        prev_delta_sum = delta_sum;
    }
}
